syntax = "proto3";

option csharp_namespace = "gRPC.telemetry";
package telemetry;

//Data stream listening on the server for client to push data into
service TelemetryService {
  rpc StreamTelemetry(stream TelemetryRequest) returns (TelemetryResponse);
}

message TelemetryRequest {
  string hostname = 1;
  string ip_address = 2;
  string uuid = 3;
  string operating_system = 4;

  oneof payload {
    NetworkStatsList network = 6;
    DiskStatsList disks = 7;
    FileSharesList shares = 8;
    DiskErrorsList disk_errors = 9;
    SystemErrorsList system_errors = 10;
    SystemUsage system_usage = 11;

  }
}

//List of all machine interfaces
message NetworkStatsList {
  repeated NetworkStats entries = 1;
}

message NetworkStats {
  string interface_name = 1;
  double rx_mbps = 2;
  double tx_mbps = 3;

  oneof interface_role {
    bool is_main = 4;
  }
}

//List of all the disks on a machine
message DiskStatsList {
  repeated DiskStats entries = 1;
}

message DiskStats {
  string mount_point = 1;
  int64 usage = 2;
  int64 capacity = 3;

}

//List of all fileshares on a machine
message FileSharesList {
  repeated FileShares entries = 1;
}

message FileShares {
  string share_path = 1;
  int64 usage = 2;
  int64 capacity = 3;
}

//List of all disk related errors
message DiskErrorsList {
  repeated DiskErrors entries = 1;
}

message DiskErrors {
  string message = 1;
  string source = 2;
  int64 timestamp = 3;
  string mount_point = 4;
}

//List of all critical machine errors
message SystemErrorsList {
  repeated SystemErrors entries = 1;
}

message SystemErrors {
  string message = 1;
  string source = 2;
  int64 timestamp = 3;
}

//List of all system usage telemetry
message SystemUsage {
  double cpu_usage_percent = 1;
  uint64 ram_total_bytes = 2;
  uint64 ram_used_bytes = 3;
  uint64 pagefile_total_bytes = 4;
  uint64 pagefile_used_bytes = 5;
  int64 last_boot_timestamp = 6;

}

//Server response
message TelemetryResponse {
  string status = 1;
}